be/src/exec/workgroup/scan_task_queue.cpp

```diff
@@ -47,168 +41,169 @@ void ScanTaskQueueWithWorkGroup::close() {
+        _update_min_wg();
     }
 
-    _total_task_num--;
+    _num_tasks--;
 
-    return wg->scan_task_queue()->take(worker_id);
+    return wg_entity->queue()->take();
 }
 
-bool ScanTaskQueueWithWorkGroup::try_offer(ScanTask task) {
+bool WorkGroupScanTaskQueue::try_offer(ScanTask task) {
     std::lock_guard<std::mutex> lock(_global_mutex);
 
-    WorkGroupPtr wg = task.workgroup;
-    wg->scan_task_queue()->try_offer(std::move(task));
-    if (_ready_wgs.find(wg) == _ready_wgs.end()) {
-        _ready_wgs.emplace(wg);
+    auto* wg_entity = task.workgroup->scan_sched_entity();
+    wg_entity->set_in_queue(this);
+    RETURN_IF_UNLIKELY(!wg_entity->queue()->try_offer(std::move(task)), false);
+
+    if (_ready_wg_entities.find(wg_entity) == _ready_wg_entities.end()) {
+        _sum_cpu_limit += wg_entity->cpu_limit();
+
+        if (auto* min_wg_entity = _min_wg_entity.load(); min_wg_entity != nullptr) {
+            // The workgroup maybe leaves for a long time, which results in that the runtime of it
+            // may be much smaller than the other workgroups. If the runtime isn't adjusted, the others
+            // will starve. Therefore, the runtime is adjusted according the minimum vruntime in _ready_wgs,
+            // and give it half of ideal runtime in a schedule period as compensation.
+            int64_t new_vruntime_ns = std::min(min_wg_entity->vruntime_ns() - _ideal_runtime_ns(wg_entity) / 2,
+                                               min_wg_entity->runtime_ns() / int64_t(wg_entity->cpu_limit()));
+            int64_t diff_vruntime_ns = new_vruntime_ns - wg_entity->vruntime_ns();
+            if (diff_vruntime_ns > 0) {
+                DCHECK(_ready_wg_entities.find(wg_entity) == _ready_wg_entities.end());
+                wg_entity->incr_runtime_ns(diff_vruntime_ns * wg_entity->cpu_limit());
+            }
+        }
+
+        _ready_wg_entities.emplace(wg_entity);
+        _update_min_wg();
     }
 
-    _total_task_num++;
+    _num_tasks++;
     _cv.notify_one();
     return true;
 }
 
-void ScanTaskQueueWithWorkGroup::_maybe_adjust_weight() {
-    if (--_remaining_schedule_num_period > 0) {
-        return;
-    }
+void WorkGroupScanTaskQueue::update_statistics(WorkGroup* wg, int64_t runtime_ns) {
+    std::lock_guard<std::mutex> lock(_global_mutex);
 
-    _cal_wg_cpu_real_use_ratio();
+    auto* wg_entity = wg->scan_sched_entity();
 
-    int num_tasks = 0;
-    // calculate all wg factors
-    for (auto& wg : _ready_wgs) {
-        wg->estimate_trend_factor_period();
-        num_tasks += wg->scan_task_queue()->size();
+    _update_bandwidth_control_period();
+    if (!wg_entity->is_rt_wg()) {
+        _bandwidth_usage_ns += runtime_ns;
     }
 
-    _remaining_schedule_num_period = std::min(MAX_SCHEDULE_NUM_PERIOD, num_tasks);
+    bool is_in_queue = _ready_wg_entities.find(wg_entity) != _ready_wg_entities.end();
+    if (is_in_queue) {
+        _ready_wg_entities.erase(wg_entity);
```

Comment:
```
    if (is_in_queue) {
        _ready_wg_entities.erase(wg_entity);
    }
    DCHECK(_ready_wg_entities.find(wg_entity) == _ready_wg_entities.end());
    wg_entity->incr_runtime_ns(runtime_ns);
    if (is_in_queue) {
        _ready_wg_entities.emplace(wg_entity);
        _update_min_wg();
    }
```
why here need erase and emplace, suggest to refine code as follows:
```cpp
auto it _ready_wg_entities.find(wg_entity);
auto in_queue = it!= _ready_wg_entities.end();
wg_entity->incr_runtime_ns(runtime_ns);
if (!in_queue) {
    _ready_wg_entities.emplace(wg_entity);
}
_update_min_wg();
```

Reply:
Because the key of _ready_wg_entities if `wg_entity->vruntime()`, we cannot modify the runtime of the entity, which is in the queue.
Therefore, we must dequeue -> modify runtime -> enqueue.